### ISSUE TITLE
Support for tuples to express per-component refness/storage

### ISSUE THREAD

**Issue by dlongnecke-cray** [chapel-lang/chapel#21717]
As a Chapel user, I want to be able to express a tuple type such as
the following where the constness and storage of each component can
vary individually as per my needs:

```chapel
type T = (const int, ref int);
```

I want to use these more flexible tuple types to better express
patterns like the Chapel standard `map` iterator, or other collection
iterators:

```chapel
record myMap {
  type K;
  type V;
  var k: K;
  var v: V;

  iter these(): (const ref K, ref V) { yield (k, v); }
}

var m: myMap(int, int);
writeln(m);               // {k = 0, v = 0}
for (k, v) in m {
//k = 4;                  <-- Cannot write, 'k' is 'const ref'
  v = 8;
}
writeln(m);               // {k = 0, v = 8}
```

**Comment by vasslitvinov** [comment]
While there, I suggest that we discuss whether referential tuples can be passed into a function by `const ref` or even `ref` intent. Analogously, whether a [const] ref of a referential tuple is legal, which is allowed by today's compiler:
```chpl
var a, b, c: ...;
ref tup = (a,b,c);
```

**Comment by dlongnecke-cray** [comment]
I'm not sure I understand how this is possible without formalizing a compiler pragma to "auto infer storage". Consider:

```chapel
record r {}
proc foo(ref tup: (int, r)) {}
```

When the type for `tup` is constructed, the storage (e.g., `ref` or `var`/`in`) of its components has to be determined at that time. It's not specified there in the syntax, so we have to choose something. It's also not generic (I don't think we have something like "auto infer storage" outside of compiler pragmas).

What we've chosen is to say that both the `int` and the `r` are stored by value. That is consistent with the rules for most things that are passed by reference, which is that they have to be some sort of `lvalue`.

So the storage of `tup` is `(in int, in r)`.

If I try to write:

```chapel
var a = 0;
var b = new r();
foo((a, b));
```

The storage of the literal `(a, b)` is `(in int, ref r)`. So we can't just pass that as is without conversion, because the formal and actual tuple types aren't the same. To convert to `(in int, in r)`, we'd have to copy the `ref r`, and I'm not sure we want to be doing copies of records just to pass by reference. Not only does it feel inconsistent with what we do for mutable references of other types, but it means that mutation of some components may not be visible for the caller (because we made temporary copies).

I'd consider how we can pass `1` to a `const ref int` to be a cheeky exception to this rule. We're really creating a `lvalue` for the literal under the hood to make this work. Maybe we can make it work, but because the formal and actual tuple types would still not be the same, there's still conversion going on. So if you pass a mutable component and expect to see any mutation reflected, it won't be:

```chapel
var x = 0;
var y = new r();
proc f(const ref tup: (int, r)) {
  x = 4;
  writeln(tup); // Would have to print (0, {})
}
f((x, y));
```

I guess this would be OK if you restricted the tuple components to literals only, though.

I'd say this problem is exactly why we're trying to move to a world where you can specify storage for each individual component. And if we can do that, I'd much rather people explicitly write:

```chapel
proc bar(tup: (const ref int, const ref r)) {}
bar((0, new r());
```

**Comment by vasslitvinov** [comment]
Right, in my mind I was thinking that the compiler DOES infer and represent per-component storage as part of the tuple type internally, so it generates correct code. It is just that these storage modifiers are invisible to the user.

Passing referential tuples by const ref is not our top priority based on user feedback and is not essential once we implement user-facing per-component annotations, as this issue proposes. However other than priority I do not see why we should not do it.

**Comment by dlongnecke-cray** [comment]
For it to be able to infer means that the storage is essentially generic, which would be a change from what we have today, right? I'm not sure if we have anything else like that other than ref-maybe-const intent for arrays.

For your example:

```chapel
var a, b, c: ...;
ref tup = (a,b,c);
```

I have a hunch that the reason why this works is because the compiler is incorrectly inferring the type of `tup` to be a referential tuple. I would honestly consider that to be a bug, and the compiler should emit an error instead. You've fixed a few similar bugs recently where referential tuple types incorrectly get assigned to variables or temporaries (e.g., for return intents).

E.g., in logs I see:

```
  val call_tmp[765619]:3*_ref(r)[1487719] "dead at end of block" "expr temp"  "maybe param" "temp"
  val tup[177889]:3*_ref(r)[1487719] "dead at end of block" "ref var"
```

Awhile back we had a discussion about whether or not `ref (int, int)` should mean `(ref int, ref int)`. This feels like it's doing that, or maybe the "ref" tag is just being incorrectly assigned, I'm not sure. But it probably shouldn't be doing this if we want it to be consistent with the semantics we've established, IMO.

**Comment by vasslitvinov** [comment]
We need to discuss semantics independent of compiler representation and what it is doing today. Also, good point about being generic over storage kind!

So, just for fun:

>  val call_tmp[765619]:3*_ref(r)[1487719] "dead at end of block" "expr temp"  "maybe param" "temp"
> val tup[177889]:3*_ref(r)[1487719] "dead at end of block" "ref var"

`tup` is made a ref to `call_tmp`, right?

No, the compiler here is not doing the conversion `ref (int, int)` --> `(ref int, ref int)` : `tup` is an honest ref to the referential tuple of the type 3*ref(r). Telltale: flag "ref var" on `tup`.
